alpheios_nemo_ui/data/templates/main/container.html

```diff
@@ -32,6 +35,40 @@
       <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
       <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
     <![endif]-->
+    <script>
+      console.log(`Head script execution has been started`)
+      if ('serviceWorker' in navigator) { // Checks if Service Workers are supported by the browser
+
+        /*
+        Register service worker only after a page is fully loaded so that the service worker registration
+        would not slow down page showing.
+         */
+        window.addEventListener('load', () => {
+          // Delay registration until document is parsed and fully loaded
+
+          console.log("Navigator online state is " + navigator.onLine)
+          if (navigator.onLine === true) {
+            console.log(`Registering a service worker ...`)
+            navigator.serviceWorker.register('/sw.js', {scope: '/'})
+              .then(reg => console.log('Service worker has been registered!', reg))
+              .catch(err => console.log('Service worker failed to register', err));
+          } else {
+            console.log(`Skipping a service worker registration because we're offline`)
+          }
```

Comment:
in the PWA prototype we had code like this:
```
onNewServiceWorker(registration, function() {
          showRefreshUI(registration);
        });
```
and the `showRefreshUI` function added a UI element for the user to accept the update or defer it.  Is there any reason we wouldn't want to carry that behavior over here?

Reply:
> Is there any reason we wouldn't want to carry that behavior over here?

That's the question of what UX would we want to present to the user. Without such a UI, when a service worker is updated, it will be loaded almost immediately, but will not be activated on pages while there are older versions of a service worker active within other browser tabs. So the activation of the new version of a service worker will be postponed until any of the following will happen:
a. User will close all tabs of the site with the PWA.
b. User will close the browser, which is a subset of (a) 🙂.
Without that, an old version of the worker will be keeping control of the open pages for a, potentially, very long time.

This is a problem in development, when service worker is updated often and we need to load those changes fast. In production, however, we probably won't update the service worker file that much and those changes, when happened, probably won't be so critical (but they might be). So it might be not important to the end user what version of the service worker is running.

One reason to introduce this UI was to allow users to force update the service worker. The other reason was to simplify development. However, with the progress of browser's dev tools the second reason is not valid any more: there are now tools in a browser that allow to force reload the service worker on each page refresh. So the only reason to have this UI is to allow users to expedite service worker updates.

Important factor in the decision is whether we do a pre-caching, and in what file is the code that loads local resources into cache. If such code is in the service worker file, we would like to force users to update it faster. But if such code is in `index.html`, then those resource will be updated even when the old service worker is in control.

The question is, then, how important would that be to users to have the fresh service worker file. On one side, having more control is good. On the other side, this may put off those who are less tech-savvy. The latter group would probably prefer everything happening invisibly behind the scenes. 

So what do you think would be an optimal way to handle this for the Alpheios Reading Tools?